src/types/router.ts

```diff
@@ -13,7 +14,10 @@ export type RouterPushOptions = {
   replace?: boolean,
 }
 
-export type RouterPush = (url: string, options?: RouterPushOptions) => Promise<void>
+export type RouterPush<TRoutes extends Routes = Routes> = {
+  (url: string, options?: RouterPushOptions): Promise<void>,
+  <TRoute extends keyof Flattened<TRoutes>>(route: { name: TRoute }, params?: Flattened<TRoutes>[TRoute], options?: RouterPushOptions): Promise<void>,
```

Comment:
I'm thinking through the devx/syntax of these three different ways of navigating to a route. 
- Route method
- Route Flat (`parent.child` & params)
- Url
I think we might want the route name to include route params and options like
```
{
  route: 'parent.child',
  params: {},
  query: {},
}
```
That way the overload becomes
```
// this type needs generics for Flattened
type RouterPushRoute = {
  route: T,
  params: TRoutes<T>,
  query: Record<string, unknown>,
  replace?: boolean,
}

// this need generics as well
type RouterPush = {
  (url: string, options?: RouterPushOptions} => Promise<void>,
  (route: RouterPushRoute) => Promise<void>
  (routeMethod => () => ReturnType<RouteMethod>) => Promise<void>
}
```
That way the arguments are unique types and we can type guard easily. 

And then we can match the same syntax with `RouterLink` where anything `router.push` accepts `RouterLink.to` can accept as well. This might not make sense when reading. Happy to chat through this.

Reply:
I don't feel strongly so I will make this change 👍